Nested catalog merge optimisation

mcb30 opened the pull request:
This is an untested first draft of the nested catalog merge optimisation described in #2707.  It is compile-tested only.

mcb30 committed:
Allow ReportModification() to truncate recursion

Change the return type of the CatalogDiffTool::ReportModification()
callback to allow an implementation of ReportModification() to
indicate that further recursion is unnecessary.  This may be used to
save substantial amounts of work by optimising away unnecessary
recursions.

No recursion is performed after calling the ReportAddition() or
ReportRemoval() callbacks, and so the return type for those callbacks
does not need to be similarly modified.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Avoid recursing into paths outside the lease when performing a merge

The existing logic will ignore any modifications reported for
directories (or files) outside of the lease path.  All files within
such a directory will also be outside the lease path, and so recursion
may be skipped to save some time.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Provide WritableCatalogManager::SwapNestedCatalog() method

Provide a method that can be used to replace the nested catalog at a
given mountpoint as a single operation.

The new nested catalog is deliberately not attached, since doing so
would cause the catalog to be recompressed and uploaded on commit.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Do not recurse into replacement nested catalogs when merging

When replacing an existing nested catalog with a new nested catalog,
there is no need for the merge operation to recurse into both old and
new nested catalogs and perform updates on an entry-by-entry basis.
The merge may be effected simply by updating the appropriate entry
within the nested_catalogs database table (and by updating the
statistics counters for the parent catalog).

When using a gateway to handle large volumes of grafted files and with
the lease paths aligned with the nested catalog mountpoints, this
optimisation can reduce the amount of work performed by the gateway to
trivial levels: instead of handling a few hundred thousand individual
directory entry catalog merge operations, the gateway needs to perform
only a single nested catalog replacement.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

jblomer commented:
Many thanks for starting this effort!  Let's have a first look at the CI.

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot unittest

jblomer commented on cvmfs/catalog_mgr_rw.cc:
```diff
@@ -56,7 +56,12 @@ WritableCatalogManager::WritableCatalogManager(
   sync_lock_ =
     reinterpret_cast<pthread_mutex_t *>(smalloc(sizeof(pthread_mutex_t)));
-  int retval = pthread_mutex_init(sync_lock_, NULL);
+  pthread_mutexattr_t attr;
```
I'm not sure why we need a recursive lock.

mcb30 replied:
The recursive lock is there to allow `WritableCatalogManager::SwapNestedCatalog` to call `WritableCatalogManager::RemoveNestedCatalog`, to avoid code duplication.  It would not be required if the code is restructured to avoid calling `WritableCatalogManager` methods.

jblomer reviewed (changes requested):
Many thanks for working on this!  I think in principal the patch takes the right approach!

The tricky bit is the implementation of `WritableCatalogManager::SwapCatalog`. We have to select carefully when to use methods from `WritableCatalogManager` and when to prefer methods from `WritableCatalog`. The `WritableCatalogManager` can attach nested catalogs behind our back, which is not what we want in this case (e.g., `WritableCatalogManager::FindCatalog()`).

I think we should use `WritableCatalogManager` methods for the parent path string to make sure all nested catalogs up the `mountpoint` but not including `mountpoint` are attached.  The rest of the swapping logic should be implemented by calls to `WritableCatalog` methods of that parent.

We could walk through the code together if you like, let me know if you want to do a shared screen session.

mcb30 commented:
> We could walk through the code together if you like, let me know if you want to do a shared screen session.

@jblomer Thank you - that would be extremely useful!  As I mentioned in https://github.com/cvmfs/cvmfs/issues/2707#issuecomment-841705500, I found it difficult to understand the expected separation of responsibilities between `WritableCatalogManager` and `WritableCatalog`, so it would be very useful to have some guidance on that.  Do you have availability tomorrow (Tuesday)?

mcb30 commented:
@jblomer I've rebased onto current `devel`, and refactored it to avoid using `WritableCatalogManager` methods (essentially by duplicating the logic from `WritableCatalogManager::RemoveNestedCatalog` within `WritableCatalogManager::SwapNestedCatalog`, so that there's no need to subsequently call `WritableCatalogManager::FindCatalog` in order to get the parent catalog object).  Hopefully this is along the lines you were thinking.

The previous version is still available within the [`mcb30:nested-merge-v1`](https://github.com/mcb30/cvmfs/tree/nested-merge-v1) branch.

mcb30 committed:
Provide LoadFreeCatalog() helper method

Provide a helper method that can be used to download a catalog file
and create a freely attached catalog object.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 commented:
@jblomer I've updated it with the new structure following this morning's discussion - thank you again!

jblomer commented on cvmfs/catalog_mgr_rw.cc:
```diff
@@ -885,6 +885,77 @@ void WritableCatalogManager::RemoveNestedCatalog(const string &mountpoint,
+  }
+
+  // Load freely attached old catalog
+  UniquePtr<Catalog> old_catalog(LoadFreeCatalog(PathString("", 0), old_hash));
```
The `mountpoint` passed to `LoadFreeCatalog` should be the nested path, I think.

mcb30 replied:
Thanks, have updated.

jblomer commented on cvmfs/catalog_mgr_rw.cc:
```diff
@@ -885,6 +885,77 @@ void WritableCatalogManager::RemoveNestedCatalog(const string &mountpoint,
+  }
+
+  // Load freely attached new catalog
+  UniquePtr<Catalog> new_catalog(LoadFreeCatalog(PathString("", 0), new_hash));
```
See above

jblomer commented on cvmfs/catalog_mgr_rw.cc:
```diff
@@ -885,6 +885,77 @@ void WritableCatalogManager::RemoveNestedCatalog(const string &mountpoint,
 
 
+/**
+ * Swap in a new nested catalog
```
Let's add a comment on the pre- / post-condition, namely that none of the nested catalogs is or gets attached to the catalog tree.

jblomer reviewed (commented):
Very well - I think that can work nicely!  Let's add some tests to the PR before merging.

jblomer commented on cvmfs/receiver/catalog_merge_tool_impl.h:
```diff
@@ -193,7 +193,24 @@ bool CatalogMergeTool<RwCatalogMgr, RoCatalogMgr>::ReportModification(
+              "CatalogMergeTool - nested catalog %s not found. Aborting",
+              rel_path.c_str());
+      }
+      output_catalog_mgr_->SwapNestedCatalog(std::string(rel_path.c_str()),
```
Simply `output_catalog_mgr_->SwapNestedCatalog(rel_path,` ?

mcb30 committed:
Add unit tests for WritableCatalogManager::SwapNestedCatalog()

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 commented:
Rebased onto current `devel` (including the merged #2718), and now includes unit tests.  Integration tests still to be done.

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cpplint

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
Test 802 seems to be unhappy

mcb30 commented:
> Test 802 seems to be unhappy

The error I'm seeing locally is
```
PANIC: /home/mcb30/work/extcvs/cvmfs/cvmfs/catalog_mgr_rw.cc : 928
failed to swap nested catalog '/a': already attached
```

I think the problem is that the call to `output_catalog_mgr_->TouchDirectory()` ends up calling `FindCatalog()` and therefore attaching the old catalog:

https://github.com/cvmfs/cvmfs/blob/65dd0082ddeb2f3ba299f59ba5016191d6698799/cvmfs/catalog_mgr_rw.cc#L703-L707

This violates the precondition that the old catalog has not been previously attached, and (correctly) triggers the PANIC error.

It feels as though `SwapNestedCatalog()` ought to be allowed to handle the situation in which the old catalog happens to be already attached, with the overall effect being similar to first calling `RemoveNestedCatalog()` with `merge=false`.

@jblomer Is there a reason why `SwapNestedCatalog()` should not be allowed to just detach the old catalog?

jblomer commented:
Oh, I see. But in this case, let's just move down the `TouchDirectory` call in the merge tool.  We only need it if we don't swap nested catalogs.

I'm hesitant to detach because in general it's hard to figure out what exactly changed in the attached subtree.  We currently want to call `SwapNestedCatalog` only in a context where we didn't make changes to the tree, but the API would still allow for misuse.

mcb30 commented:
> Oh, I see. But in this case, let's just move down the `TouchDirectory` call in the merge tool. We only need it if we don't swap nested catalogs.

Done.

I also had to modify the code to use
```c++
const bool found = new_catalog_mgr->LookupNested(path, ...)
```
rather than
```c++
const bool found = new_catalog_mgr->LookupNested(rel_path, ...)
```
It seems as though `MakeRelative()` strips off any leading slash, while `MakeRelativePath()` inserts a leading slash. :confused:

@jblomer Test 802 should now pass, and turns out to already test the ability to swap a nested catalog (that contains another nested catalog).  Do we still need a separate integration test?

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cpplint

jblomer commented:
@mcb30 Good point on the integration tests. Let me have a look.  We might indeed already be well covered.

jblomer commented:
@mcb30 The integration tests look good!  I'll take a final look over the code today.

jblomer commented on test/unittests/t_catalog_mgr_rw.cc:
```diff
@@ -118,4 +131,98 @@ TEST_F(T_CatalogMgrRw, CloneTree) {
+}
+
+
+TEST_F(T_CatalogMgrRw, SwapNestedCatalogFailSlow) {
```
This test should not take longer than 1-2 seconds, does it? The `EXPECT_DEATH` are expensive but there are no network timeouts as far as I can see.  Let's make it a regular test then so that it runs routinely.

```suggestion
TEST_F(T_CatalogMgrRw, SwapNestedCatalogFail) {
```

mcb30 replied:
It seems to take about 10 seconds.  I based the choice upon the existing `CloneTreeFailSlow` test within the same file, which AFAICT is "slow" only in the sense of using `EXPECT_DEATH()`.

I'll make the change - let me know if it should be "slow" after all.

jblomer replied:
Oh, it takes _that_ long.  Apologies, then the `Slow` suffix is of course appropriate.

jblomer commented on test/unittests/t_catalog_mgr_rw.cc:
```diff
@@ -118,4 +131,98 @@ TEST_F(T_CatalogMgrRw, CloneTree) {
+
+  // Fail if parent catalog does not exist
+  catalog_mgr->DetachNested();
+  EXPECT_DEATH(catalog_mgr->SwapNestedCatalog("/dir/dir/dir/sub1",
```
It's currently subtle: the nested catalog does not exist due to the leading slash, right?

```suggestion
  EXPECT_DEATH(catalog_mgr->SwapNestedCatalog("/no/such/directory/sub1",
```

mcb30 replied:
You're right: I think this is a relic from when I was still being surprised by the unexpected behaviours of `MakeRelative()`/`MakeRelativePath()` (see https://github.com/cvmfs/cvmfs/pull/2709#issuecomment-853102465).

Will update.

jblomer reviewed (changes requested):
I think we are almost there.  Two tiny suggestions, and in addition I think it would be good to slightly extend integration test 802.  Between transaction 2 and transaction 3, it would be good to exercise some changes, e.g. adding one or two files to /a and /a/b, ideally in such a way that the counts of files change wrt. the previous state.

Really cool improvement!

mcb30 commented:
> I think we are almost there. Two tiny suggestions, and in addition I think it would be good to slightly extend integration test 802. Between transaction 2 and transaction 3, it would be good to exercise some changes, e.g. adding one or two files to /a and /a/b, ideally in such a way that the counts of files change wrt. the previous state.

Have updated as per the suggestions, and I'll add the extension of test 802 tomorrow.

> Really cool improvement!

Thank you, and thanks for all of your help with this change!

mcb30 committed:
Extend gateway integration tests to include nested catalog file additions

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Actually check for test failures in repository gateway nested catalogs

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 commented:
Having updated test 802 to actually check for test failures, I'm seeing errors of the form:
```
linkcounts differ: 2 / 3
timestamps differ: 1622820061 / 1622820074
transition point and root entry differ (/a)
```
I assume the timestamp difference arises because something needs to explicitly update the timestamp as recorded in the parent catalog to match the swapped-in nested catalog?

mcb30 commented:
> I assume the timestamp difference arises because something needs to explicitly update the timestamp as recorded in the parent catalog to match the swapped-in nested catalog?

As far as I can tell, all of the existing code that handles this kind of thing does so by mounting the nested catalog and updating the root directory entry in the nested catalog to match the transition point directory entry in the parent catalog.  Since we're trying to avoid touching the new nested catalog, I'm assuming that we should do basically the reverse: get the root directory entry from the freely attached new catalog object, and use that to update the transition point directory entry in the parent catalog?

@jblomer Is this the correct approach?

jblomer commented:
Good catch!  Yes, we forgot about updating the transition point in the parent catalog entry. It should work as you describe.

mcb30 commented:
> Good catch! Yes, we forgot about updating the transition point in the parent catalog entry. It should work as you describe.

Updated - `SwapNestedCatalog()` now looks up the root directory entry from the freely attached new catalog object, and applies it using:
```c++
  dirent.set_is_nested_catalog_mountpoint(true);
  dirent.set_is_nested_catalog_root(false);
  parent->UpdateEntry(dirent, nested_root_path);
  parent->TouchEntry(dirent, xattrs, nested_root_path);
```
It seems to be necessary to use both `UpdateEntry()` and `TouchEntry()` - AFAICT the former is required to update the link count and the latter is required to update the extended attributes.

This now passes the updated test 802 and I have verified that the test code really is checking counters etc (by temporarily hacking out the call to `PopulateToParent()`).

@jblomer I believe this should now be ready to merge.

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cpplint
